flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultPartitionManager.java

```diff
@@ -113,10 +126,12 @@ void onConsumedPartition(ResultPartition partition) {
 			final ResultPartition previous = registeredPartitions.remove(partition.getPartitionId());
 			// Release the partition if it was successfully removed
 			if (partition == previous) {
-				partition.release();
-				ResultPartitionID partitionId = partition.getPartitionId();
-				LOG.debug("Released partition {} produced by {}.",
-					partitionId.getPartitionId(), partitionId.getProducerId());
+				if (!partition.isManagedExternally() || isReleaseExternallyManagedPartitionsOnConsumption) {
```

Comment:
this isn't sufficient; the partition is already removed from `registeredPartitions` above, leading to wrong behavior of `getUnreleasedPartitions`

Reply:
Yes, I also noticed this just now, the state is inconsistent here if removed but not released.